### PR TITLE
Adding js/ts util method to verify JWT's generated by a scanner

### PR DESCRIPTION
**Pull request by Robert-H-Leonard**
### Feature work
- [x] Added new method to JS/TS sdk to verify a JWT generated by a scanner
- [x] Added new method to python sdk to verify a JWT generated by a scanner

### Testing
**Manual/Local testing**

In order to validate the verification locally I generated a JWT locally and called the verify method during the initialization of my bot running locally. B/c this bot isn't deployed to validate the happy path locally, during testing I found a bot and an scanner who were linked and hardcoded their addresses in the call to `areTheyLinked` to make sure the smart contract call was valid:

- BotId: 0xe0889c513f9b4fcad87f9c26e9df2e25474223b3b77f7348ee5398bec4e0ee73
- ScannerId: 0xBF536900A88F13D02bc1ff80Af84e480c591001E

### CODE REVIEW

File: sdk/utils.ts [hunk 1]
```diff
@@ -188,6 +189,66 @@ export const fetchJwtToken = async (claims: {}, expiresAt?: Date): Promise<{toke
...
+
+  if(!botId) {
+    console.warn(`Invalid claim`)
+    return false
+  }
+
+  const publicKey = payload?.sub as string | undefined // public key should be contract address that signed the JWT
+
+  if(!publicKey) {
+    console.warn(`Invalid claim`)
```
Comment (Robert-H-Leonard) [inline sdk/utils.ts:236]: Question: Not sure how much info we want to say when we detect invalid claims

**Comment by Robert-H-Leonard** [comment]
In terms of E2E testing are there any other ways to test without having to manually deploy a bot?

**Review by haseebrabbani — COMMENTED** [review]
i'm not seeing the Python SDK changes in the PR (maybe forgot to commit?)

**Comment by haseebrabbani** [comment]
> In terms of E2E testing are there any other ways to test without having to manually deploy a bot?

E2E would imply that we are testing the whole stack (and therefore deploying a bot). we can definitely unit test these methods (just created a ticket to add unit tests for SDK methods https://github.com/forta-network/forta-bot-sdk/issues/213)

File: sdk/utils.ts [hunk 2]
```diff
@@ -188,6 +189,66 @@ export const fetchJwtToken = async (claims: {}, expiresAt?: Date): Promise<{toke
...
+  const rawHeader = splitJwt[0]
+  const rawPayload = splitJwt[1]
+
+  const header = JSON.parse(Buffer.from(rawHeader, 'base64').toString())
+  const payload = JSON.parse(Buffer.from(rawPayload, 'base64').toString())
+
+  const botId = payload["bot-id"] as string
+  const algorithm = header?.alg;
+
+  if(algorithm !== "ETH") {
```
[the remark this answers is not in the document]
Comment (Robert-H-Leonard) [reply sdk/utils.ts:223]: I was copying the logic here: https://github.com/forta-network/forta-core-go/blob/master/security/jwt.go#L84

pinging Caner to follow up.

File: sdk/utils.ts [hunk 3]
```diff
@@ -188,6 +189,66 @@ export const fetchJwtToken = async (claims: {}, expiresAt?: Date): Promise<{toke
...
+    console.warn(`Unexpected signing method: ${algorithm}`)
+    return false
+  }
+
+  if(!botId) {
+    console.warn(`Invalid claim`)
+    return false
+  }
+
+  const publicKey = payload?.sub as string | undefined // public key should be contract address that signed the JWT
```
[the remark this answers is not in the document]
Comment (Robert-H-Leonard) [reply sdk/utils.ts]: Updated

File: sdk/utils.ts [hunk 4]
```diff
@@ -188,6 +189,66 @@ export const fetchJwtToken = async (claims: {}, expiresAt?: Date): Promise<{toke
...
+  const signature = `0x${ Buffer.from(splitJwt[2], 'base64').toString('hex')}`
+
+  const signerAddress = ethers.utils.recoverAddress(digest, signature) // Contract address that signed message
+
+  if(signerAddress !== publicKey) {
+    console.warn(`Signature invalid: expected=${publicKey}, got=${signerAddress}`)
+    return false
+  }
+
+  const dispatchContract = new ethers.Contract(DISPATCH_CONTRACT, [DISPTACHER_ARE_THEY_LINKED], getEthersProvider())
```
[the remark this answers is not in the document]
Comment (Robert-H-Leonard) [reply sdk/utils.ts]: Updated to always use a polygon provider.

File: python-sdk/src/forta_agent/utils.py [hunk 5]
```diff
@@ -130,7 +130,7 @@ def keccak256(val):
     hash.update(bytes(val, encoding='utf-8'))
     return f'0x{hash.hexdigest()}'
 
-def fetch_Jwt_token(claims, expiresAt=None) -> str:
+def fetch_Jwt(claims, expiresAt=None) -> str:
```
Comment (haseebrabbani) [inline python-sdk/src/forta_agent/utils.py:133]: fetch_jwt*

File: sdk/utils.ts [hunk 6]
```diff
@@ -188,6 +193,66 @@ export const fetchJwtToken = async (claims: {}, expiresAt?: Date): Promise<{toke
...
+  if(algorithm !== "ETH") {
+    console.warn(`Unexpected signing method: ${algorithm}`)
+    return false
+  }
+
+  if(!botId) {
+    console.warn(`Invalid claim`)
+    return false
+  }
+
```
Comment (haseebrabbani) [inline sdk/utils.ts:236]: we also need to verify token expiry timestamp somewhere here

File: sdk/utils.ts [hunk 7]
```diff
@@ -188,6 +193,66 @@ export const fetchJwtToken = async (claims: {}, expiresAt?: Date): Promise<{toke
...
+  return {
+    header,
+    payload
+  }
+}
+
+const DISPTACHER_ARE_THEY_LINKED = "function areTheyLinked(uint256 agentId, uint256 scannerId) external view returns(bool)";
+const DISPATCH_CONTRACT = "0xd46832F3f8EA8bDEFe5316696c0364F01b31a573"; // Source: https://docs.forta.network/en/latest/smart-contracts/
+
+export const verifyJwt = async (token: string): Promise<boolean> => {
```
[the remark this answers is not in the document]
Comment (Robert-H-Leonard) [reply sdk/utils.ts]: Updated.

File: sdk/utils.ts (hunk 6 above)
Comment (Robert-H-Leonard) [reply sdk/utils.ts:233]: Added a check for the expiry time

File: sdk/utils.ts [hunk 8]
```diff
@@ -21,6 +22,10 @@ export const getEthersBatchProvider = () => {
   return new ethers.providers.JsonRpcBatchProvider(getJsonRpcUrl())
 }
 
+const getPolygonProvider = () => {
```
[the remark this answers is not in the document]
Comment (Robert-H-Leonard) [reply sdk/utils.ts]: Removed

File: python-sdk/src/forta_agent/utils.py [hunk 9]
```diff
@@ -163,6 +163,6 @@ def fetch_Jwt_token(claims, expiresAt=None) -> str:
             raise err
 
 
-def decode_Jwt_token(token):
+def decode_Jwt(token):
     # Adding need 4 byte for pythons b64decode
     return base64.b64decode(token.split('.')[1] + '==').decode('utf-8')
```
[the remark this answers is not in the document]
Comment (Robert-H-Leonard) [reply python-sdk/src/forta_agent/utils.py]: Updated to use urlsafe_b64

File: python-sdk/src/forta_agent/__init__.py [hunk 10]
```diff
@@ -7,7 +7,7 @@
 from .trace import Trace, TraceAction, TraceResult
 from .event_type import EventType
 from .network import Network
-from .utils import get_json_rpc_url, create_block_event, create_transaction_event, get_web3_provider, keccak256, get_transaction_receipt, get_alerts, fetch_Jwt_token, decode_Jwt_token
+from .utils import get_json_rpc_url, create_block_event, create_transaction_event, get_web3_provider, keccak256, get_transaction_receipt, get_alerts, fetch_Jwt, decode_Jwt
```
[the remark this answers is not in the document]
Comment (Robert-H-Leonard) [reply python-sdk/src/forta_agent/__init__.py]: Updated

File: python-sdk/src/forta_agent/utils.py [hunk 11]
```diff
@@ -163,6 +163,6 @@ def fetch_Jwt_token(claims, expiresAt=None) -> str:
             raise err
 
 
-def decode_Jwt_token(token):
+def decode_Jwt(token):
```
[the remark this answers is not in the document]
Comment (Robert-H-Leonard) [reply python-sdk/src/forta_agent/utils.py]: Updated

File: python-sdk/src/forta_agent/utils.py (hunk 5 above)
Comment (Robert-H-Leonard) [reply python-sdk/src/forta_agent/utils.py]: Updated

**Review by haseebrabbani — APPROVED** [review]

**Pull request merged**